dev/Kozani/KozaniRemoteManager/KozaniDvcServer.cpp

```diff
@@ -0,0 +1,734 @@
+        // that size and sent/received as multiple operations.
+        BYTE readBuffer[CHANNEL_PDU_LENGTH];
+        std::vector<BYTE> pduBuffer;
+        UINT32 pduSize{};
+        bool isPartialData{};
+        OVERLAPPED overlappedRead{};
+        overlappedRead.hEvent = readComplete.get();
+
+        while (!exitThread)
+        {
+            DWORD error{};
+
+            // Set up async read on the DVC.
+            if (!ReadFile(m_dvcHandle.get(), readBuffer, sizeof(readBuffer), nullptr, &overlappedRead)
+                && ((error = GetLastError()) != ERROR_IO_PENDING))
+            {
+                hr = HRESULT_FROM_WIN32(error);
+                DvcListenerThreadExiting(hr, "[DVC] read failure");
+                RETURN_HR(hr);
+            }
+
+            HANDLE events[] = { m_dvcThreadExit.get(), readComplete.get() };
+            LogDebugMessage("[DVC] Listener thread starting to wait");
+
+            // When we reach this point, the DVC listener thread has passed setup phase and is running.
+            m_dvcThreadStarted.SetEvent();
+
+            DWORD result = WaitForMultipleObjects(ARRAYSIZE(events), events, FALSE, INFINITE);
+            LogDebugMessage("[DVC] Listener thread wait completed (result=0x%x)", result);
+            if (result == WAIT_FAILED)
+            {
+                hr = HRESULT_FROM_WIN32(GetLastError());
+                DvcListenerThreadExiting(hr, "[DVC] WaitForMultipleObjects failure");
+                RETURN_HR(hr);
+            }
+
+            switch (result)
+            {
+                case WAIT_OBJECT_0:
+                    // Thread exit handle signaled.
+                    LogDebugMessage("[DVC] Thread exit event handled");
+                    return S_OK;
+
+                case WAIT_OBJECT_0 + 1:
+                {
+                    // DVC read completion.
+                    DWORD bytesRead{};
+                    if (GetOverlappedResult(m_dvcHandle.get(), &overlappedRead, &bytesRead, FALSE))
+                    {
+                        const BYTE* pdu{};
+                        // Failing to process DVC data will not bubble up the failure, which will end the DVC listener thread. 
+                        // As a server, it should be robust to mal-formatted data.
+                        if (SUCCEEDED_LOG(ProcessDvcDataChunk(readBuffer, bytesRead, isPartialData, pduBuffer, pduSize, &pdu)))
+                        {
+                            LOG_IF_FAILED_MSG(ProcessProtocolDataUnit(pdu, pduSize), "Failed to process PDU. Ignore failure.");
+                        }
+                    }
+                    else
+                    {
+                        hr = HRESULT_FROM_WIN32(GetLastError());
+                        DvcListenerThreadExiting(hr, "[DVC] GetOverlappedResult() failed");
+                        RETURN_HR(hr);
```

Comment:
DvcListenerThreadExiting() itself can throw and if we use CATCH_RETURN() MACRO it will hide the original error, which is the one we want to return. If we simply use try/catch(...) then we lose the exception to HRESULT conversion from CATCH_RETURN() MACRO. Currently DvcListenerThreadExiting does not return a result and any exception from it will be converter to HRESULT and log (using CATCH_LOG_RETURN). It does not affect the original HRESULT we want to return to the caller.